### ISSUE TITLE
Hang or Deadlock when calling ray.get() inside pytorch Dataset when DataLoader with num_workers >0

### ISSUE THREAD

**Issue by jamesliu** [ray-project/ray#13407]
Ray
### What is the problem?
When num_workers > 0 in PyTorch DataLoader, the ray.get hangs forever. It is ok when num_workers =0
*Ray version and other system information (Python version, TensorFlow version, OS):*
I tested in Ubuntu, Python 3.7, Pytorch 1.7.0
### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):
```
import torch
import ray
from torch.utils.data.dataset import IterableDataset
from torch.utils.data.dataloader import DataLoader
import numpy
from itertools import islice

@ray.remote
class ReplayBuffer():
    def __init__(self, batch_size = 5):
        self.data = numpy.arange(0, 10)
        self.batch_size = batch_size           
    def get_random_batch(self):
        idx = numpy.random.choice(len(self.data) - self.batch_size)
        return self.data[idx:idx+self.batch_size]
    
ray.init(ignore_reinit_error=True)
replay_buffer_actor = ReplayBuffer.remote()

class MyDataset(IterableDataset):
    def __init__(self, replay_buffer_actor):
        self.replay_buffer_actor = replay_buffer_actor
    def __iter__(self):
        while True:
            yield ray.get(self.replay_buffer_actor.get_random_batch.remote())

dataset = MyDataset(replay_buffer_actor)
dataloader = DataLoader(dataset, num_workers = 1)
max_steps = 3
for batch in islice(dataloader, max_steps):
    print(batch)
```
If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [X] I have verified my script runs in a clean environment and reproduces the issue.
- [X] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by rkooo567** [comment]
cc @richardliaw

**Comment by faroit** [comment]
I wonder if this is a similar issue when using joblib for local parallelization: https://github.com/facebookresearch/hydra/issues/964

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
